streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -63,83 +87,184 @@ public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient,
+                "This can happen if the Kafka cluster is temporary not available. " +
+                "You can increase admin client config `retries` to be resilient against this error.";
+            log.error(timeoutAndRetryError);
+            throw new StreamsException(timeoutAndRetryError);
         }
-        throw new StreamsException("Could not create internal topics.");
     }
 
     /**
      * Get the number of partitions for the given topics
      */
     public Map<String, Integer> getNumPartitions(final Set<String> topics) {
-        for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
-            try {
-                final MetadataResponse metadata = streamsKafkaClient.fetchMetadata();
-                final Map<String, Integer> existingTopicPartitions = fetchExistingPartitionCountByTopic(metadata);
-                existingTopicPartitions.keySet().retainAll(topics);
-
-                return existingTopicPartitions;
-            } catch (StreamsException ex) {
-                log.warn("Could not get number of partitions: {} Retry #{}", ex.getMessage(), i);
+        return getNumPartitions(topics, false);
+    }
+
+    private Map<String, Integer> getNumPartitions(final Set<String> topics,
+                                                  final boolean bestEffort) {
+        int remainingAttempts = 1 + retries;
+        boolean retry;
+        while (remainingAttempts > 0) {
+            retry = false;
+
+            final DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(topics);
+            final Map<String, KafkaFuture<TopicDescription>> futures = describeTopicsResult.values();
+
+            final Map<String, Integer> existingNumberOfPartitionsPerTopic = new HashMap<>();
+            for (final Map.Entry<String, KafkaFuture<TopicDescription>> topicFuture : futures.entrySet()) {
+                try {
+                    final TopicDescription topicDescription = topicFuture.getValue().get();
+                    existingNumberOfPartitionsPerTopic.put(
+                        topicFuture.getKey(),
+                        topicDescription.partitions().size());
+                } catch (final InterruptedException fatalException) {
+                    Thread.currentThread().interrupt();
+                    log.error("Thread got interrupted.", fatalException);
+                    throw new StreamsException(fatalException);
+                } catch (final ExecutionException couldNotDescribeTopicException) {
+                    final Throwable cause = couldNotDescribeTopicException.getCause();
+                    if (cause instanceof TimeoutException) {
+                        retry = true;
+                        log.debug("Could not get number of partitions for topic {} due to timeout. " +
+                            "Will try again (remaining retries {}).", topicFuture.getKey(), remainingAttempts - 1);
+                        try {
```

Comment:
I think we do not need to back off here, since the request will be parked in the queue anyways during retries.